### PR TITLE
RUST-1358 Remove most type constraints on cursor values

### PR DESCRIPTION
**Pull request by abr-egn**
RUST-1358

Now that we're internally passing around `RawDocument` rather than the deserialized type, we don't need to carry around the final type for our internal cursor implementation bits, and in turn we don't need the `Sync + Send + Unpin` constraints on that type.

I think there's further opportunity for cleaning up the implementation, but that won't have any user-facing benefit, so I've filed RUST-1676 to track that.

### CODE REVIEW

**Review by isabelatkinson — APPROVED** [review]
looks good, just one question!

File: src/cursor/mod.rs [hunk 1]
```diff
@@ -104,11 +104,11 @@ pub struct Cursor<T> {
     client: Client,
     // `wrapped_cursor` is an `Option` so that it can be `None` for the `drop` impl for a cursor
     // that's had `with_type` called; in all other circumstances it will be `Some`.
-    wrapped_cursor: Option<ImplicitSessionCursor<T>>,
+    wrapped_cursor: Option<ImplicitSessionCursor>,
     drop_address: Option<ServerAddress>,
     #[cfg(test)]
     kill_watcher: Option<oneshot::Sender<()>>,
-    _phantom: std::marker::PhantomData<T>,
+    _phantom: std::marker::PhantomData<fn() -> T>,
```
[the remark this answers is not in the document]
Comment (abr-egn) [reply src/cursor/mod.rs:111]: This is a gnarly little corner of the Rust typesystem: https://doc.rust-lang.org/nomicon/phantom-data.html

TL;DR changing it from `PhantomData<T>` to `PhantomData<fn() -> T>` means the compiler will no longer consider dropping `Cursor<T>` to drop a `T`, which can relax restrictions in some hypothetical situations.
